### ISSUE TITLE
bug(ffi/cuda): cuda_launch4i kernels silently stop executing after many prior launches (and huge-grid params-buffer kernels truncate) — blocks Qwen3 tied-lm_head

### ISSUE THREAD

**Issue by rickenator** [rickenator/Vyn#210]
**bug(ffi/cuda): `cuda_launch4i` kernels silently stop executing after many prior launches (and huge-grid params-buffer kernels truncate) — blocks the Qwen3 tied-lm_head**

## Summary
In Vyb kernel-mode host code (RTX 3090, sm_86, Vyb 0.7.3 Debug, `~/Projects/Vyb/build/vyb`), kernels launched via the `cuda_launch4i` freedom shim (a thin `cuLaunchKernel` wrapper) silently **stop doing work after a large number of prior launches in the same process**, even though every call returns rc=0. Also, a params-buffer kernel launched with a very large grid (1.5M blocks) executes only a leading fraction.

## Evidence (real Qwen3-4B prefill driver)
1. **Per-layer dequant + forward works across 36 layers** — ~1100 consecutive `cuda_launch4i` launches (q4kdeq/q6kdeq up to **97,280 blocks** in one launch, gemm/rmsnorm/rope/attn/silu/resid) all execute, and the full-36-layer hidden matches a numpy reference to maxrel ~5e-6. So ordinary launches and large single launches work.
2. **A long sequential loop of small launches stops executing**: `emb_dequant` dequantizes the 151,936×2560 `token_embd` (Q6_K) in 5935 chunks, each launched as `cuda_launch4i(q6kdeq, CH/256=256, 1,1, 256, ...)` (256 blocks). Every call returns 0, but after the first ~dozens of chunks the kernels stop writing: `DE[0]` and `DE[2560]` (rows 0,1) are correct but `DE[row2000]=0` and `DE[row100000]=0` (both should be nonzero). So ~5900 launches are dropped silently.
3. **The SAME `emb_dequant` works standalone**: a probe that runs only this dequant (few prior launches) fills `DE` fully and the lm_head argmax is exactly correct (31784, matching numpy). So it is not the kernel or the dequant logic — it is an interaction with the number of prior launches in the process.
4. **Huge-grid params-buffer kernel truncates**: a `trans_emb` launch of `D*VOCAB/256 ≈ 1,519,360` blocks executes only the leading ~170 blocks (wrote embedding rows 0..~16, rest zero), silently. (Note: layer `gemm`/`rmsnorm` params-buffer kernels with grids ≤ ~1187 blocks and up to 76–256 blocks DO execute fully, so the cutoff is not "params buffers are broken".)

## Impact
Any Vyb program doing many chained GPU kernel launches with real data beyond a modest count, or one very-wide grid, silently under-computes — the rest of the grid reads as untouched (0) memory. This blocks the Qwen3-4B tied-lm_head (logits = hidden @ token_embdᵀ over 151,936 vocab) for the VybForge native-inference pipeline.

## Repro sketch
```
for i in 0..6000:
    chunk = read 53KB packed token_embd slice
    cuMemcpyHtoD(dpk, chunk, nbytes)
    rc = cuda_launch4i(q6kdeq, 256 blocks, ..., dpk, DE + i*65536*8, 65536, 0)  # always rc=0
# observe: DE beyond ~row 26 is all 0
```

## Suspected root cause
Some bounded resource/limit in the Vyb `cuda_launch4i` path (queued launches being dropped, or a gridDim/launch-budget cap that is not surfaced as an rc error), plus a separate large-gridDim truncation for params-buffer kernels. The wrapper at `src/main.cpp:1884` calls `cuLaunchKernel` directly and returns its rc (0 here) — the drop is not reported.

## Ask
- Surface launch failures as real error codes (don't silently no-op).
- Investigate the launch-count / gridDim limits in the kernel-mode launch path (likely the `%ctaid`/launch parameters for the Vyb kernel-mode lowering) and fix the silent under-execution.
- Regression: loop `cuda_launch4i` 6000× on distinct output slices and assert all are written.

**Comment by rickenator** [comment]
Investigated on the same RTX 3090 / sm_86 toolchain. I was unable to reproduce
either symptom through the Vyb compiler launch path, and my strongest probe
argues **cuda_launch4i is not silently dropping launches**.

**Many-launch probe (`fill` kernel x 6000):** allocated one device buffer, kicked
6000 `cuda_launch4i(fill, 1,1,1, 1,1,1, D + i*4, 0,0,0)` launches with NO
intermediate sync, then a single `cuCtxSynchronize` + readback:
- rc != 0 count: **0**
- dropped slots (read != 42): **0** (all 6000 written)

So with the repo's simple fixtures the identical many-launch pattern executes
fully. The kernel-mode grid reads use proper NVPTX `${ctaid}`/`${nctaid}`
intrinsics and gridX passes through `cuda_launch4i` as an `unsigned` with no cap,
so 1,519,360 blocks is not truncated by Vyb codegen.

**What this points to (VybForge driver side, not rickenator/Vyb):**
1. The "emb_dequant first ~dozens chunks correct, later 0" signature is exactly a
   **read-before-async-queue-drains** race: if the driver kicks all 5935 launches
   (which reuse `dpk` via HtoD) and reads `DE[i]` back WITHOUT a per-chunk
   `cuCtxSynchronize`, the leading ~dozens have executed and the rest read as
   untouched 0. The standalone probe "works" because it runs few launches that
   fit in the queue before its readback.
2. The huge-grid `trans_emb` truncation (~170 leading blocks) is in that kernel's
   own grid/params config, not the compiler — a single 1.5M-block launch is legal
   CUDA (gridDimX max 2^31-1) and is passed through verbatim.

**Concrete next steps for VybForge (with the compiler unblocking you):**
- After each chunk loop (or before each readback group), call `cuCtxSynchronize()`
  and CHECK its return — a non-zero value surfaces the async launch error that
  `cuLaunchKernel`'s rc=0 hides (your "silently no-op" ask is exactly this: the
  compiler's shims return the enqueue status; only a sync error can reveal an
  enqueue-accepted/execute-failed launch).
- For `trans_emb`, log its actual gridDimX/blockDimX and confirm the kernel idx =
  blockIdx.x*blockDim.x+threadIdx.x stays in range; verify the params buffer length
  matches the kernel's expected params.

I'm leaving this open (not closed) since it did not reproduce as a Vyb compiler
bug here. If, after the sync-check above, launches still drop on the real Qwen3
driver, post the `cuCtxSynchronize` return and the first dropped offset and I'll
re-open a compiler angle.

**Comment by rickenator** [comment]
Resolved — NOT a Vyb bug, closing. Root cause was in MY VybForge driver, confirmed & fixed: in the token_embd dequant loop I wrote `freedom{ ... return 0 }` inside the loop body; that `return 0` exits the whole function after the FIRST chunk, so only ~26 embedding rows were dequantized (rest of DE stayed 0). That made the downstream transpose/gemm look 'truncated'. Fixed by capturing the error into a function-scope var and only returning after the loop; all chunk launches now execute and the lm_head argmax exactly matches the numpy reference (2-layer: 85926/30935; full-36 pending). Apologies for the noise — nothing wrong with cuda_launch4i.
